### ISSUE TITLE
About LASSO based channel pruning

### ISSUE THREAD

**Issue by JinyangGuo** [NervanaSystems/distiller#109]
Hi there,

Thanks for open-source the code. 

Is there any plan for implementation of the LASSO based channel-pruning algorithm (i.e. the paper: Channel pruning for accelerating very deep neural networks)?

**Comment by nzmora** [comment]
Hi,

The above paper reports the following ResNet50 results:

![image](https://user-images.githubusercontent.com/31280975/50097684-14f19b80-0223-11e9-8571-fc36af75e845.png)

Distiller currently has an implementation of filter pruning based on filter L1-norm + AGP pruning rate control, which achieves 2.24x compute reduction and with 0.87% Top5 reduction (the pruning schedule is [here](https://github.com/NervanaSystems/distiller/blob/master/examples/agp-pruning/resnet50.schedule_agp.filters.yaml)).  

This is only one comparison point, but it the moment it is enough to make this paper less attractive.
[Discrimination-aware Channel Pruning for Deep Neural Networks](https://arxiv.org/abs/1810.11809) shows some of the best results I've seen so far, so it is more interesting.  But I cannot tell you when we will have an implementation (you are welcome to go for it!).

Cheers,
Neta

**Comment by zheLim** [comment]
> Hi,
> 
> The above paper reports the following ResNet50 results:
> 
> ![image](https://user-images.githubusercontent.com/31280975/50097684-14f19b80-0223-11e9-8571-fc36af75e845.png)
> 
> Distiller currently has an implementation of filter pruning based on filter L1-norm + AGP pruning rate control, which achieves 2.24x compute reduction and with 0.87% Top5 reduction (the pruning schedule is [here](https://github.com/NervanaSystems/distiller/blob/master/examples/agp-pruning/resnet50.schedule_agp.filters.yaml)).
> 
> This is only one comparison point, but it the moment it is enough to make this paper less attractive.
> [Discrimination-aware Channel Pruning for Deep Neural Networks](https://arxiv.org/abs/1810.11809) shows some of the best results I've seen so far, so it is more interesting. But I cannot tell you when we will have an implementation (you are welcome to go for it!).
> 
> Cheers,
> Neta

This paper "Discrimination-aware Channel Pruning for Deep Neural Networks" can only apply to classification task. Since discrimination loss requires labels. 
I think lasso based channel pruning has better generalization ability. Since its loss is not related to target loss( such as classification loss, objectness loss).

**Comment by nzmora** [comment]
Hi @zheLim ,

We would love it if you chose to implement and share [Channel Pruning for Accelerating Very Deep Neural Networks](https://arxiv.org/pdf/1707.06168.pdf).
Please let us know if you need any help.

Thanks!
Neta

**Comment by zheLim** [comment]
> Hi @zheLim ,
> 
> We would love it if you chose to implement and share [Channel Pruning for Accelerating Very Deep Neural Networks](https://arxiv.org/pdf/1707.06168.pdf).
> Please let us know if you need any help.
> 
> Thanks!
> Neta

I would like to try. But it seems that current working pipeline of distiller does not fit  lasso regression well. Lasso regression based method prunes channel layer by layer and do fine-tuning after all layer pruning are done. But heuristic based method prunes all layer once and train on every epoch. How to integrate lasso regression based method with heuristic based method is a problem.

**Comment by levzlotnik** [comment]
Hi,

Closing due to staleness.
